designs/expose-metrics.md

```diff
@@ -0,0 +1,371 @@
+* capc-controller-manager controller
+* capv-controller-manager controller
+* capx-controller-manager controller
+
+**Out of Scope:**
+
+Following components are not considered for exposing metrics securely:
+
+* Snow provider (capas) and docker provider (capd)
+* ECR-credential-provider
+
+**Future Scope:**
+
+* Securely expose metrics for all other components (kube-vip, coredns, cilium, cert-manager)
+
+## Current State of EKS-A components
+
+![table](images/expose-metrics.png)
+
+## Overview of Solution
+
+There are two general solutions proposed for different components here:
+
+* For k8s system components, documenting the steps to configure some proxy as a daemonset. No code changes needed in EKS-A
+* For all other controller-runtime based CAPI and EKS-A components, implementing the CAPI [diagnostics](https://main.cluster-api.sigs.k8s.io/tasks/diagnostics) feature and exposing the bind address with 0.0.0.0
+
+For k8s system components, we don't just want to change the default bind address to 0.0.0.0 because configuring it to bind on all interfaces might expose the metrics publicly over the internet on a node which has any interface exposed to the internet even if that component has authentication and authorization enabled. (Check this [issue](https://github.com/kubernetes/kubeadm/issues/2244#issuecomment-763294722) for more details). Also, it goes against the principle of security [hardening](https://en.wikipedia.org/wiki/Hardening_(computing)) where the default configuration should be kept minimal to reduce the attack surface of the system. For all other controller-runtime based components, it is best to implement the diagnostics feature that CAPI has introduced to match the core controllers and also have consistency across all these components. It also removes the current dependency we have on the [kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy) for capc controller. Overall, there will be no API changes in the cluster spec for any of the components.
+
+#### **Kube-Apiserver and Kubelet:**
+
+These components already perform bearer token authentication and RBAC authorization for client requests and they are already configured to allow listening on all interfaces and IP address families. No further action needs to be taken to expose metrics for these components securely.
+
+#### **Kube-Controller-Manager, Kube-Scheduler and Kube-Proxy:**
+
+Kube-controller-manager and kube-scheduler already perform bearer token authentication and RBAC authorization whereas kube-proxy does not but all three components listen only on the localhost (127.0.0.1) for client requests. We can document the steps to be followed to configure some proxy as a Daemonset on the cluster which forwards the client requests to the metrics endpoint for each component. The proxy pods must run in the `hostNetwork` so that they can access the loopback interfaces of the corresponding pods. There will be no changes done in EKS-A to configure these for the customers.
```

Comment:
```suggestion
Kube-controller-manager and kube-scheduler already perform bearer token authentication and RBAC authorization, whereas kube-proxy does not. But all three components listen only on the localhost (127.0.0.1) for client requests. We can document the steps to be followed to configure some proxy as a Daemonset on the cluster, which forwards the client requests to the metrics endpoint for each component. The proxy pods must run in the `hostNetwork` so that they can access the loopback interfaces of the corresponding pods. There will be no changes done in EKS Anywhere to configure these for the customers.
```